csrc/id_model/id_model.h

```diff
@@ -64,6 +64,19 @@ StatefulInliningInfo buildStatefulInliningInfo(
 // considered the exact same size operating on matching dimensions from the root
 // domain mapping.
 //
+// LOOP mode is important to resolve inlined broadcassts. If we have something
+// like: consumer[i0o, threadIdx.x{i0i}] = producer[i0o,
+// threadIdx.y{i0i}](computeAt = 1) which can easily happen when using shared
+// memory. Loop is actually defined for all iteration domains, and resembles
+// groups of iter domains that are effectively inlined with each other.
+// Therefore iter domain's that are a common dependency of inlined leaf domains
+// may be loop mapped together.
```

Comment:
I would be very interested in seeing when the loop promotion we have is strictly required, and when it is just one way to do things. For example, if I have a fusion
```
T0[1, 4]
T1[3, 4] = T0[1, 4]

T1->reorder({{0, 1}});
T1->merge(0);
T1->split(2, inner=false);
propagate;
T0->inlineAt(1);
```
Then for this specific case, loop promotion is not necessary, because `(i, j)` in the leaf domain of T1 is `((i*6+j)%3, (i*6+j)/3)` in T1's root domain. For T0, without loop promotion, `(i, k)` in the leaf domain is `(0, i*2+k)` in T0's root domain. According to Theorem 2.15.1 in https://github.com/NVIDIA/Fuser/blob/main/doc/math/integer-division.md, `(i*6+j)/3 = i*2+j/3`, which has very similar mathematical form as `i*2+k`. And this mathematical similarity tells us that each `i` takes the same slice of T0 and T1, therefore the program is valid even without loop promotion.

Reply:
That's very interesting. 🤯